src/Shared/BinarySerializer.cs

```diff
@@ -4,43 +4,79 @@
 //
 
 using System.IO;
-using System.Runtime.Serialization.Formatters.Binary;
+using System.Text;
+using System.Web.SessionState;
+using ProtoBuf;
 
 namespace Microsoft.Web.Redis
 {
+
     public class BinarySerializer : ISerializer
     {
+        private readonly byte redisNull = 1;
+        private readonly byte initializeItem = 2;
+        private readonly byte str = 3;
         public byte[] Serialize(object data)
         {
-            if (data == null)
-            {
-                data = new RedisNull();
-            }
-            var binaryFormatter = new BinaryFormatter();
             using (var memoryStream = new MemoryStream())
             {
-                binaryFormatter.Serialize(memoryStream, data);
-                byte[] objectDataAsStream = memoryStream.ToArray();
-                return objectDataAsStream;
+                if (data is null)
+                {
+                    return new byte[] { redisNull };
+                }
+                if (data is SessionStateActions.InitializeItem)
+                {
+                    return new byte[] { initializeItem };
+                }
+                if (data is string)
+                {
+                    memoryStream.WriteByte(str);
+                }
+
+                Serializer.Serialize(memoryStream, data);
+                return memoryStream.ToArray();
             }
+
         }
 
         public object Deserialize(byte[] data)
         {
-            if (data == null)
+            if (data is null)
+            {
+                return null;
+            }
+
+            byte firstByte = data[0];
+
+            if (firstByte == redisNull)
             {
                 return null;
             }
-            var binaryFormatter = new BinaryFormatter();
-            using (var memoryStream = new MemoryStream(data, 0, data.Length))
+
+            if (firstByte == initializeItem)
             {
-                memoryStream.Seek(0, SeekOrigin.Begin);
-                object retObject = (object)binaryFormatter.Deserialize(memoryStream);
-                if (retObject.GetType() == typeof(RedisNull))
+                return SessionStateActions.InitializeItem;
+            }
+
+            using (var memoryStream = new MemoryStream(data))
+            {
+                if (firstByte == str)
+                {
+                    memoryStream.ReadByte();
+                }
+
+                byte[] retObject = Serializer.Deserialize<byte[]>(memoryStream);
+
+
+                if (firstByte == str)
                 {
-                    return null;
+                    return Encoding.Default.GetString(retObject);
```

Comment:
do not use `Encoding.Default` (ever); this property is either:

- completely undefined, invariable wrong, and not unicode-compatible (.net framework etc)
- the same thing as utf8 (.net core etc)

In all cases, I suspect you should probably be using UTF8 here